test/test_qubit_spectroscopy.py

```diff
@@ -0,0 +1,139 @@
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Spectroscopy tests."""
+
+from typing import Tuple
+
+import numpy as np
+from qiskit.qobj.utils import MeasLevel
+from qiskit.test import QiskitTestCase
+
+from qiskit_experiments.characterization.qubit_spectroscopy import QubitSpectroscopy
+from qiskit_experiments.test.mock_iq_backend import TestJob, IQTestBackend
+
+
+class SpectroscopyBackend(IQTestBackend):
+    """
+    A simple and primitive backend to test spectroscopy experiments.
+    """
+
+    def __init__(
+        self,
+        line_width: float = 2e6,
+        freq_offset: float = 0.0,
+        iq_cluster_centers: Tuple[float, float, float, float] = (1.0, 1.0, -1.0, -1.0),
+        iq_cluster_width: float = 0.2,
+    ):
+        """Initialize the spectroscopy backend."""
+
+        self.__configuration__["basis_gates"] = ["spec"]
+
+        self._linewidth = line_width
+        self._freq_offset = freq_offset
+
+        super().__init__(iq_cluster_centers, iq_cluster_width)
+
+    # pylint: disable = arguments-differ
+    def run(self, circuits, shots=1024, meas_level=MeasLevel.KERNELED, **options):
+        """Run the spectroscopy backend."""
+
+        result = {
+            "backend_name": "spectroscopy backend",
+            "backend_version": "0",
+            "qobj_id": 0,
+            "job_id": 0,
+            "success": True,
+            "results": [],
+        }
+
+        for circ in circuits:
+
+            run_result = {
+                "shots": shots,
+                "success": True,
+                "header": {"metadata": circ.metadata},
+            }
+
+            set_freq = float(circ.data[0][0].params[0])
+            delta_freq = set_freq - self._freq_offset
+            prob = np.exp(-(delta_freq ** 2) / (2 * self._linewidth ** 2))
+
+            if meas_level == MeasLevel.CLASSIFIED:
+                counts = {"1": 0, "0": 0}
+
+                for _ in range(shots):
+                    counts[str(np.random.binomial(1, prob))] += 1
+
+                run_result["data"] = {"counts": counts}
+            else:
+                memory = [self._draw_iq_shot(prob) for _ in range(shots)]
+                run_result["data"] = {"memory": memory}
+
+            result["results"].append(run_result)
+
+        return TestJob(self, result)
+
+
+class TestQubitSpectroscopy(QiskitTestCase):
+    """Test spectroscopy experiment."""
+
+    def setUp(self):
+        """Setup."""
+        super().setUp()
+        np.random.seed(seed=10)
+
+    def test_spectroscopy_end2end_classified(self):
```

Comment:
I'm not sure if we really need such integration test. I'm fine with adding test for `.circuit` and `.analyis_run`. How do we guarantee if the test backend works correctly when we run into the trouble?

Reply:
Testing the circuit and the analysis_run is probably not enough. Testing the circuit is almost a trivial test. Testing the analysis will tend to end up checking only the fitting which in itself is already good but not enough IMO.